toolbox/include/base/Parameters.h

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (C) 2018 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * GNU Lesser General Public License v2.1 or any later version.
+ */
+
+#ifndef WBT_PARAMETERS_H
+#define WBT_PARAMETERS_H
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace wbt {
+    template <typename T>
+    class Parameter;
+    class ParameterMetadata;
+    class Parameters;
+    const int PARAM_INVALID_INDEX = -1;
+    const std::string PARAM_INVALID_NAME = {};
+} // namespace wbt
+
+/**
+ * @brief Class for storing block's parameters
+ *
+ * This class can contain scalar and vector parameters of the supported types.
+ *
+ * @see wbt::Parameter, wbt::ParameterMetadata, wbt::ParameterType
+ */
+class wbt::Parameters
+{
+public:
+    using ParamIndex = int;
+    using ParamName = std::string;
+
+private:
+    class impl;
+    std::unique_ptr<impl, void (*)(impl*)> pImpl;
```

Comment:
I am confused here. You are passing `void (*)(impl*)` as the deleter, and I am not sure that I understand what this mean at the C++ level to be honest. Who is actually deleting the class? Are you sure that this is not creating a memory leak?

Reply:
Sorry, I got it during the review. `void (*)(impl*)` is the type of the deleter, and the actual deleter is passed in the unique_ptr constructor. I think you mentioned me f2f that the need for doing this is due to the necessity of defining a copy constructor, it would be nice to have a small commenter here hinting that.

Reply:
Yes something like that. If I understood correctly, the issue here is use usage of:

```cpp
~WBBlock() override = default;
```

I used as an example `WBBlock` which is **not** copiable.

The combination of the default destructor plus the forward declaration of the object handled by the `unique_pointer` makes `sizeof` complain. `sizeof` is used in the default deleter, for this reason passing a custom deleter solves this issue.

Reply:
> The combination of the default destructor plus the forward declaration of the object handled by the unique_pointer makes sizeof complain. sizeof is used in the default deleter, for this reason passing a custom deleter solves this issue.

The combination of the default destructor plus the forward declaration of the object handled by the unique_pointer is a standard feature of the pimpl pattern, see for example all the usage in the blocks such as https://github.com/robotology/wb-toolbox/pull/77/files#diff-d7d137bec79209b2e9a3a4548fd95e51R25, and there seems not to be problem with `sizeof` .

If I understood correctly, the problem is combining this with a non-copyable class?  In that case it is not clear to me how it is related to sizeof, but I do not strongly feel about this issue, so if it is working for you it is fine for me.

Reply:
From [unique_ptr](http://en.cppreference.com/w/cpp/memory/unique_ptr):

> std::unique_ptr may be constructed for an incomplete type T, such as to facilitate the use as a handle in the pImpl idiom. If the default deleter is used, T must be complete at the point in code where the deleter is invoked, which happens in the destructor, move assignment operator, and reset member function of std::unique_ptr.

As discusses f2f, when possible removing the custom deleter and moving the default destructor to the cpp is preferred for readability.